mpi-proxy-split/mpi_plugin.cpp

```diff
@@ -461,17 +553,40 @@ mpi_plugin_event_hook(DmtcpEvent_t event, DmtcpEventData_t *data)
   switch (event) {
     case DMTCP_EVENT_INIT: {
       JTRACE("*** DMTCP_EVENT_INIT");
+      JASSERT(dmtcp_get_real_tid != NULL);
+      initialize_signal_handlers();
       initialize_segv_handler();
       JASSERT(!splitProcess()).Text("Failed to create, initialize lower haf");
       seq_num_init();
       mana_state = RUNNING;
+
+      DmtcpMutexInit(&g_upper_half_fsbase_lock, DMTCP_MUTEX_LLL);
+
       break;
     }
     case DMTCP_EVENT_EXIT: {
       JTRACE("*** DMTCP_EVENT_EXIT");
       seq_num_destroy();
       break;
     }
+
+    case DMTCP_EVENT_PTHREAD_START: {
+      // Do we need a mutex here?
+      DmtcpMutexLock(&g_upper_half_fsbase_lock);
```

Comment:
tid would be unique for each thread. So, I guess we can skip the mutex here.

Reply:
STL `unordered_map` is not thread-safe so that's why we need a mutex.

Reply:
I believe different threads can access and modify different elements concurrently. In our case, unique TIDs should be enough to ensure that no two threads will access the same key.

Reply:
They can access it but if one thread tries to modify unordered_map (insert, erase), that'll cause a race condition and the results are unexpected. In multi-threaded application, STL containers need to be protected with mutex.

Reply:
Got it.